compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -4586,6 +4586,47 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             _ => "/* value */".to_string(),
         })
     }
+
+    fn suggest_add_result_as_return_type(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        err: &mut Diag<'_>,
+        trait_ref: ty::PolyTraitRef<'tcx>,
+    ) {
+        if ObligationCauseCode::QuestionMark != *obligation.cause.code().peel_derives() {
+            return;
+        }
+
+        let node = self.tcx.hir_node_by_def_id(obligation.cause.body_id);
+        if let hir::Node::Item(item) = node
+            && let hir::ItemKind::Fn(sig, _, body_id) = item.kind
+            && let hir::FnRetTy::DefaultReturn(ret_span) = sig.decl.output
+            && self.tcx.is_diagnostic_item(sym::FromResidual, trait_ref.def_id())
+            && let ty::Tuple(l) = trait_ref.skip_binder().args.type_at(0).kind()
+            && l.len() == 0
+            && let ty::Adt(def, _) = trait_ref.skip_binder().args.type_at(1).kind()
+            && self.tcx.is_diagnostic_item(sym::Result, def.did())
+        {
+            let body = self.tcx.hir().body(body_id);
+            let mut sugg_spans =
+                vec![(ret_span, " -> Result<(), Box<dyn std::error::Error>>".to_string())];
+
+            if let hir::ExprKind::Block(b, _) = body.value.kind
+                && b.expr.is_none()
+            {
+                // The span will point to the closing curly brace `}` of the block.
+                sugg_spans.push((
+                    b.span.shrink_to_hi().with_lo(b.span.hi() - BytePos(1)),
```

Comment:
since we know there must be at least one statement, you can instead do `b.stmts.last().unwrap().span.shrink_to_hi()`, this way you'll get a span right after the last statement.

Reply:
> since we know there must be at least one statement, you can instead do `b.stmts.last().unwrap().span.shrink_to_hi()`, this way you'll get a span right after the last statement.

Thank you. I tried this way. But when the last stmt is a macro call like  `println!();`, it's span will point to std 
 like: `D:\source\rust\rust\library\std\src\macros.rs:85:6: 85:6 (#8)`.

Reply:
ah indeed. hmm... Maybe we should change the value of `Block::span` to not include the brackets around it. We can already obtain the "with brackets" span by taking the span of the outer expression. That's a bit more involved though, as you'd probably need to touch the parser.

I don't think the `- BytePos(1)` will work in case the block itself is created by a macro or other expansion. Please add some tests for async functions, which should show us any problematic behaviour. And try to generate an entire function with a macro and see how your suggestion behaves on that.

Reply:
> ah indeed. hmm... Maybe we should change the value of `Block::span` to not include the brackets around it. We can already obtain the "with brackets" span by taking the span of the outer expression. That's a bit more involved though, as you'd probably need to touch the parser.
> 
> I don't think the `- BytePos(1)` will work in case the block itself is created by a macro or other expansion. Please add some tests for async functions, which should show us any problematic behaviour. And try to generate an entire function with a macro and see how your suggestion behaves on that.

Hello, thank you very much.
 I tried async and declaring macros and it handles the case of declaring macros correctly.
For async functions, now the current logic will skip it. I added the corresponding conditions and check a async function like: 

```rust
async fn test2() {
    let mut file = File::create("foo.txt")?;
    println!();
}
```

The fix result will be :
```rust
async fn test2() -> Result<(), Box<dyn std::error::Error>> {
    Ok(())
}
```

The span seems point to the body correct. But after fixed the expr `let mut file = File::create("foo.txt")?;` and `println!();` in the body will be overwriten. 

Because the current logic does not handle the closure situation, and I have not thought of an effective way to repair the closure, I think we can skip it in this PR.